### ISSUE TITLE
4chan X has become unusably slow this past week (Chrome Extension)

### ISSUE THREAD

**Issue by yejip79190** [ccd0/4chan-x#2613]
For the record, I have not had this issue with any other site or any other script/extension. It ONLY happens on 4chan and ONLY when X is turned on.
Pages will hang for up to a full minute, sometimes more. At times when I open the catalog, it will just be a blank screen for about 10 seconds before anything loads. Even still, when it does load I can't interact with anything for another long period of time and no hover features work. By the time it actually does load and become functional, when I click on a thread that takes an additionally obnoxious long amount of time before it loads that. I've been having to drag threads into new tabs to make them not take forever to load. I've tried switching off the extension version and just using the userscript with TamperMonkey but that didn't work.

Speaking of threads, whenever I first enter a thread things seem to load pretty well but once a new post comes in, or whether I watch the thread, it begins to lag just like the catalog page does. When I try to reply, it'll be fine right up until the point I actually click "post" and at that point, everything freezes for like 20 seconds before I even get a notification at the top about it being successfully posted. Then throw in more time before the post even shows up or the page unfreeze. Additionally to all of that, I've noticed in the Chrome task manager that any open tab of 4chan now takes an ungodly amount of CPU usage. I never really looked at it before so maybe it's nothing new, but I thought it would be noteworthy.

I've tried three separate things that helped:
-Reset 4chan X settings
-Cleared thread watcher manually from the actual JSON file
-Disabled JSON index

At first, I thought it was just the thread watcher but I don't see how that would be possible as just as little as a week ago everything was fine. My thread watcher IS pretty big, there's over hundreds of threads in it, but again I've only just started to have problems with it and there's only been like 3 threads I've watched/replied to that were even added to it since then. For the record it has made 4chan X very slightly laggier than if I had fresh settings, but that minimal lag was nothing compared to this. I just don't see how it could be the root cause. The JSON index disable fix was the most significant fix as even with my absolutely stacked thread watcher it still made everything lag-free. But obviously a lot of those JSON index features are really critical to 4chan X and I'd like to avoid turning them off if possible. 

Another reason I feel it can't be related to the thread watcher, despite its size, is that I've even imported old settings files that I've done backups for. I tried one from well over two months ago and the problem still persisted.

Have there been any changes that could have caused this? It genuinely does make things unusable as you can't even go from one page to the other without enormous lag.

**Comment by yejip79190** [comment]
I've just tried it on Firefox and it worked like a charm, zero issues. This is definitely a Chrome thing.

**Comment by ccd0** [comment]
What version number of Chrome?

**Comment by yejip79190** [comment]
> What version number of Chrome?

Version 83.0.4103.61 (Official Build) (64-bit)

**Comment by ccd0** [comment]
Are you using any other extensions?

**Comment by yejip79190** [comment]
> Are you using any other extensions?

The only other extensions I have are uBlock Origin, a Gmail notification thing and Tampermonkey. I've tried disabling all of them and it changed nothing. I also tried a fresh installation of chrome with nothing from the previous installation, and as soon as I import the old 4chan X settings with the packed thread watcher it starts doing it again.

Maybe an update with Chrome is affecting the performance hit of the thread watcher, somehow? Because I know for a 100% fact that it wasn't doing this a week ago.

**Comment by ccd0** [comment]
I imported a few hundred threads and didn't notice anything. It might only happen with a particular setting you're using. Would you mind posting a copy of your exported settings? (Make sure to edit out any sensitive information like tripcode passwords.)

**Comment by yejip79190** [comment]
> I imported a few hundred threads and didn't notice anything. It might only happen with a particular setting you're using. Would you mind posting a copy of your exported settings? (Make sure to edit out any sensitive information like tripcode passwords.)

[Here.](https://files.catbox.moe/v1t57m.json)

I just tested these settings on another computer with a fresh Chrome profile and only 4chan X installed and it did the same thing. I also noticed that posting a reply is by far the most laggy thing. It took literally 3 minutes for a page to unfreeze itself after I posted something.

**Comment by roboxa** [comment]
I've been having the EXACT same issues.
> I've just tried it on Firefox and it worked like a charm, zero issues. This is definitely a Chrome thing.
Omg thanks for pointing this out. I tried it and it does indeed work. It's not an ideal fix but at least I can actually use 4chan on firefox! My guess is there's some kind of weird performance hangup going on with both the latest versions of chrome and 4chan X. Whatever it is, it's probably related to the watcher since I went into the file for the config settings and removed all the watched threads and it made my 4chan normal again (on chrome). But I don't need to do that for firefox at all strangely, it just works. Maybe google messed something up with how resources are allocated to extensions?

**Comment by yejip79190** [comment]
Something of interest, I switched to the Tampermonkey userscript version and downgraded to 1.14.19.0 and it seems to be working well.

**Comment by ccd0** [comment]
And do you see the same problem with 1.14.19.3 as a userscript? Does 1.14.19.2 work for you?

**Comment by yejip79190** [comment]
> And do you see the same problem with 1.14.19.3 as a userscript? Does 1.14.19.2 work for you?

Yep, I just gave it another go and 1.14.19.3 is as laggy as ever. I tried 1.14.19.2 and it's not as bad. It seems like it may still be a little laggy but I might just be paranoid. Still better than 19.3 either way.

**Comment by ccd0** [comment]
Are you using the prefetch button?

**Comment by yejip79190** [comment]
Not as far as I'm aware.  Does it only show up if the "Image Prefacing" option is ticked? I've always had that turned off.

**Comment by ccd0** [comment]
> Does it only show up if the "Image Prefetching" option is ticked?

That's correct. For you there should be no difference between 1.14.19.0, 1.14.19.2, and 1.14.19.3.

**Comment by yejip79190** [comment]
> That's correct. For you there should be no difference between 1.14.19.0, 1.14.19.2, and 1.14.19.3.

Really odd. Have you been able to replicate anything on your end?

**Comment by ccd0** [comment]
Nothing so far.

**Comment by yejip79190** [comment]
> Nothing so far.

Have you tried on my settings? For perspective, when 4chan is working correctly I always have 3 open tabs of it. One for /v/, one for /tv/ and one for /lit/. Never really had a performance problem doing it this way.  Anyway, in any one of those tabs for those boards I'll either usually just click a thread and go into it within the tab or, if I'm starting to accumulate a lot of watched active threads on that particular day, I'll start opening them in new tabs so I can monitor them without making the catalog reset and potentially purge any threads that have died in the meantime. And my tabs are almost always open on the catalog. 

So, in doing this, on the latest version of 4chan X when a tab has been open on the catalog for a good long while, I'll come back to it and instead of it just loading the tab as per usual it'll be stuck loading on a blank screen for about 10 seconds minimum. It's also REALLY noticeable when you reply to any given thread, as everything just comes to a complete stop and it's like the site is moving in slow motion. First it freezes on the captcha completion, then it freezes on posting the reply, and then the whole thread just stays frozen for what feels like minutes until you see the notification that the reply has successfully been posted. Then it's frozen for even a little bit more after that before the page finally updates and shows your post. At that point the performance will be somewhat normal, although another person replying can slow it down too. And just to be safe, I did check and make sure that it's not having multiple tabs/instances open that caused it, and all this stuff did happen again on just a single tab.

Another thing I tried that made it really noticeable is watching multiple threads in a short amount of time, like going through them on the catalog and alt-clicking so they're watched. And clicking to go to another board from the top left.

_(Sidenote: is it possible to keep the catalog from auto-updating? I recall a long time ago I used to be able to click a thread from the catalog and hit back then the catalog would look the same way as I left it, but some time ago it was updated to auto-update once you go to another page. It's not a huge deal or anything, just curious if you can change that?)_

**Comment by ccd0** [comment]
> Have you tried on my settings?

Yes, I'm just not seeing it.

Another thing that might help is if you check whether you can reproduce this on a fresh profile:
https://developer.chrome.com/devtools/docs/clean-testing-environment

> (Sidenote: is it possible to keep the catalog from auto-updating? I recall a long time ago I used to be able to click a thread from the catalog and hit back then the catalog would look the same way as I left it, but some time ago it was updated to auto-update once you go to another page. It's not a huge deal or anything, just curious if you can change that?)

I think it works this way in Firefox but not Chrome.

**Comment by yejip79190** [comment]
I've done it on a fresh profile on a completely different computer with the only shared thing being the same settings file for 4chan X, and it's still the same. Is it possible that the file or thread watcher part of it is corrupted? I have no idea why the newest version caused this or how it could've caused it.

**Comment by yejip79190** [comment]
Test

**Comment by yejip79190** [comment]
Some news, on the other computer I've tested it on I isolated it to happening most often with the extension. It still happened with the latest version of the userscript but the extension was far more consistently broken and slow. I even tried an older version of the extension and it did the same thing, so I believe this to be some issue a Chrome update has brought that affects the extension.

Right now my "solution" is just staying on 1.14.18.0 with the userscript. Alternatively, other people can use Firefox as it seems to be 100% functional at all times on Firefox regardless of what you use or what you do.

**Comment by saxamaphone69** [comment]
FWIW, testing on Chrome Version 83.0.4103.97 (Official Build) (64-bit) and 4chan X 1.14.19.3 (as an extension or with TamperMonkey + LATEST) works without issues, but I admit I do not have 100s of threads watched.

When you tried it on a fresh profile _without_ your old 4chan X settings, did you experience any lag?

You could try the Performance tab in the DevTools for Chrome. Hit record, try to recreate the steps, and you should be able to see what is causing the lag (make sure Memory is ticked towards the top). You should be able to save that snapshot too and upload it.

![image](https://user-images.githubusercontent.com/877480/84116178-afc2e300-aa72-11ea-8401-83e3a1c4a1ba.png)

**Comment by ccd0** [comment]
I think I must have mixed up your settings with #2618 when I said I had tested with your settings. When importing your settings, I do get lag. By my count you have 14946 threads in your thread watcher. There is definitely some room for efficiency improvements in the thread watcher, but you are probably hitting the limit of what 4chan X can currently handle efficiently on your machine.

**Comment by yejip79190** [comment]
> I think I must have mixed up your settings with #2618 when I said I had tested with your settings. When importing your settings, I do get lag. By my count you have 14946 threads in your thread watcher. There is definitely some room for efficiency improvements in the thread watcher, but you are probably hitting the limit of what 4chan X can currently handle efficiently on your machine.

I'm not completely against the idea of this, but how would we be able to explain it becoming an issue only recently? Like I said, when I first started having this issue my thread watcher list wasn't any bigger than it was before besides maybe 5 threads at most, which with a number as big as 14946 doesn't sound like it would make a difference. And other people have been seeming to have issues too but I don't know how big their watcher lists are.

**Comment by Redsona45** [comment]
okay SO. ive read this thread but i just want to mention anybody using a VPN browser extension (even if it's not active, just enabled via extentions manager in ANY electron browser) WILL slow down the scripts in tampermonkey/violentmonkey  (cant speak for any other script managers as i have not tried them), in turn making 4chan lag alot on using 4cx ^^ i hope this helps someone!
